README.md
# 中文 Zhongwen - A Chinese-English Pop-Up Dictionary

This browser extension allows you to translate Chinese characters and words
by simply hovering over them with the mouse.

**Chrome extension:**
https://chrome.google.com/extensions/detail/kkmlkkjojmombglmlpbpapmhcaljjkde

**Firefox extension:**
https://addons.mozilla.org/en-US/firefox/addon/zhongwen-chinese-english/

**Github:**
https://github.com/cschiller/zhongwen

## Highlights
- Supports both traditional and simplified characters.
- Includes a recent version of the widely used CEDICT Chinese English
  dictionary.
- Displays Hanyu Pinyin along both the simplified and traditional characters,
  however only showing either simplified or traditional characters
  can be configured as an option.
- As a learning aid it uses different colors for displaying the Pinyin
  syllables, depending on the tone of the Chinese character.
- Can be turned on and off with a single mouse-click.
- Highlights the characters whose translation is displayed in the pop-up
  window.
- Also supports keyboard navigation for translating the next character, the
  next word, or the previous character.
- Allows you to add words to a built-in word list. Words from this list can be
  exported to a text file for further processing, such as importing the words
  into Anki.
- Includes links to grammar and usage notes on the Chinese Grammar Wiki
  (https://resources.allsetlearning.com/chinese/grammar)
- Supports exporting words to the Skritter vocabulary queue
  (https://skritter.com/)


## Installation

The Zhongwen extension is available as Chrome and Firefox Add-on:
https://chrome.google.com/extensions/detail/kkmlkkjojmombglmlpbpapmhcaljjkde
https://addons.mozilla.org/en-US/firefox/addon/zhongwen-chinese-english/

Simply click on the "Add to Chrome" or "Add to Firefox" button respectively
to start the installation.


## How Does It Work?
Once Zhongwen is installed on your computer you'll see a little icon showing
the Zhongwen logo near the upper right-hand corner of your browser. By
clicking on this icon you can turn the extension on and off. Click on it and
wait until you see the word "On" displayed in red and white on top of the icon.
This tells you that the dictionary is active. If you want to turn it off again,
just click on the icon again and the red "On" label disappears. So with a
single mouse-click you can activate or deactivate the dictionary.

Once Zhongwen has been turned on, showing the red "On" label on the icon, you
can go to a web page in Chinese and point your mouse at some Chinese characters.
A pop-up window opens up automatically showing you the translation of the word
you were pointing at.

Zhongwen can translate both simplified characters as well as traditional
characters.


## What Information Does It Show?
By default, the pop-up window shows the following pieces of information:

- the Chinese charaters forming the word you're pointing at with your mouse. If the
  simplified and the traditional forms differ, then both forms are shown.
- the corresponding Pinyin. The individual Pinyin syllables are shown in different
  colors (see below).
- the English translation. Sometimes, Zhongwen finds more than one matching entry 
  in its dictionary. In this case it shows you all of the matching entries.


## Zhuyin
If you want to see Zhuyin, a.k.a. Bopomofo, you can enable it via the options
page (see below).


## Chinese Grammar Wiki
Zhongwen also allows you to learn more about Chinese grammar and usage. If
Zhongwen knows that there's an entry at the Chinese Grammar Wiki for the entry
you're just looking at, it will show you the following text at the bottom
of the pop-up window: 'Press "g" for grammar and usage notes.' If you then
press "g" on your keyboard, a new tab will open in your browser and it will
take you directly to the corresponding page in the Chinese Grammar Wiki.


## Why Is Pinyin Shown in Different Colors?
If you're learning Chinese, Zhongwen helps you remember the tone of each
character by displaying it in a different color. It uses the following color
scheme for showing Pinying syllables:

- First tone syllables are shown in red.
- Second tone syllables are shown in orange.
- Third tone syllables are shown in green.
- Forth tone syllables are shown in blue.


## Keyboard Navigation
Some users don't use Zhongwen to look up individual words, they want to read
an entire paragraph or article. In this case, rather than having to trace all
the words with the mouse, you can use Zhongwen's built-in keyboard navigation
support: You can move from character to character or from word to word by
pressing keys on your keyboard:

- Pressing "n" on the keyboard takes you to the next word.
- Pressing "b" takes you back one character.
- Pressing "m" moves to the next character. (This is different from pressing "n"
  if you're currently over a word that consists of two or more characters.)
- Sometimes the pop-up window is hiding the text you're looking at. You can move
  it out of the way using "x" and "y" on the keyboard (this moves the pop-up
  window up or down), or hit "a" once or twice to move it to an altogether
  different location.


## Built-in Word List
So now that you can look up all those words when reading Chinese web sites, wouldn't
it be nice to add the new words you want to remember to some list that you can use for
studying? Zhongwen let's you do that, too! It has a built- in word list feature. You just
need to press "r" (short for "remember") on your keyboard and the entries you're looking
at in the currently opened pop-up window will be added to the built-in word list. How
do you get to see this list? It's easy, you need to use press the "Alt" and the "w" key
on your keyboard (Alt + w). The word list will then open in a new tab.


## Can I export the dictionary entries into a spreadsheet or import them into Anki?

Yes, if you want to copy the current dictionary entry, including the Chinese characters,
Pinyin, and the English translation, to the clipboard, just hit "c" on the keyboard to
copy to the clipboard. (Note: This is different from using "Ctrl + C" on a PC or
"Command + C" on a Mac, which only copies the Chinese characters without the translation.)
When you paste the clipboard into Excel or an OpenOffice spreadsheet, the individual
components of the entry are nicely put in different columns and rows. From there you can
also import them into Anki. (See the Anki manual for further details.)


## Skritter Support
If you're studying Chinese characters using Skritter you can easily add Chinese words
to your vocabulary queue: just hit "s" on the keyboard. This opens up a new tab that
takes you to the Skritter web site. If you're not logged into Skritter you'll see the
login page. After you're logged in, the word is automatically added to your study queue.


## Context Menu Items

When Zhongwen is enabled, you can right-click on a page to open the browser's
context menu. There will be a menu item for Zhogwen which you can use to access
a new tab showing a helpful summary of all the keyboard shortcuts. Another
menu item can be used for opening the word list as an alternative to the
"Alt + w" keyboard shortcut.


## What Built-In Dictionary Does It Use?
Zhongwen comes with a recent release of the CC-CEDICT dictionary. This dictionary is
constantly updated, and each Zhongwen release will include an updated version of the
dictionary.


## Support for Online Dictionaries
Sometimes, in addition to looking at the translation in the pop-up window, you might
want to look up a word in one of the freely available online dictionaries. Zhongwen
lets you quickly do that by some simple keyboard shortcuts.

* "Alt + 1" looks up the selected word using LINE Dict.
* "Alt + 3" takes you to Dict.cn.
* "Alt + 4" uses iCIBA.
* "Alt + 5" looks up the word in the MDBG dictionary.
* "Alt + 6" takes you to JuKuu.
* "t" is for looking up example sentences at Tatoeba.org

Your browser will open a new tab and the word that was shown in the pop-up window
will be looked up in the corresponding online dictionary listed above.


## More Options
Zhongwen also has an options page which allows you to configure certain aspects
of its behavior. In order to get there just right-click on the Zhongwen logo
in the upper right-hand corner of the browser and choose "Options". A new tab
will open showing you all the available options with a brief explanation of
each one.


## Legal
This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

---

*Copyright (C) 2019 Christian Schiller*


Changes.md
# Change Log
## 4.19.1
- Allow displaying only simplified or traditional characters instead of both in
  the popup (fixes #22, thanks to @CNurt)
- Fix display of zhuyin with umlauts (fixes #20, thanks to @CNurt)

## 4.19
- Update CC-CEDICT dictionary

## 4.18
- Update CC-CEDICT dictionary

## 4.17
- Update CC-CEDICT dictionary

## 4.16
- Update CC-CEDICT dictionary

## 4.15
- Update CC-CEDICT dictionary

## 4.14.2
- Option to toggle Zhongwen via keyboard shortcut

## 4.14.1
- Fix unwanted keypress interception by Zhongwen in rare cases (#10)

## 4.14
- Update CC-CEDICT dictionary

## 4.13
- Update CC-CEDICT dictionary

## 4.12
- Update CC-CEDICT dictionary

## 4.11.5
- Really fix for addon activation on tab change / new tabs

## 4.11.4
- Fix key handling

## 4.11.3
- Fix copy to clipboard

## 4.11.2
- Fix display of '</span>' for zhuyin

## 4.11.1
- Bug fixes for addon enabling

## 4.11
- Update CC-CEDICT dictionary to version 2017-12-10
- Disable yellowbridge support due to changes on their site

## 4.9.7
- Show wordlist context menu entry even if popup dictionary is not active, but
  only in the browser_action menu and not the page menu.

## 4.9.6
- Preserve addon activation status between browser restarts
- Show wordlist context menu entry even if popup dictionary is not active

## 4.9.5
- Fix wordlist saving and deletion
- Update jqgrid to 4.15.2

## 4.9.4
- Support 'Preferences' button for about:addons
  - Settings are synced between different computers
- Refactor background scripts to use more asynchronous operations

## Older
- Migration of deprecated Chrome (and for Firefox unimplemented) APIs.
- Code refactoring to pass Mozilla's QA
  - Remove unneeded permissions
  - Update 3rd party libraries
  - Use document fragments to construct popup content


dict.js
/*
        Zhongwen - A Chinese-English Popup Dictionary
        Copyright (C) 2011 Christian Schiller
        https://chrome.google.com/extensions/detail/kkmlkkjojmombglmlpbpapmhcaljjkde

        ---

        Originally based on Rikaikun 0.8
        Copyright (C) 2010 Erek Speed
        http://code.google.com/p/rikaikun/

        ---

        Originally based on Rikaichan 1.07
        by Jonathan Zarate
        http://www.polarcloud.com/

        ---

        Originally based on RikaiXUL 0.4 by Todd Rudick
        http://www.rikai.com/
        http://rikaixul.mozdev.org/

        ---

        This program is free software; you can redistribute it and/or modify
        it under the terms of the GNU General Public License as published by
        the Free Software Foundation; either version 2 of the License, or
        (at your option) any later version.

        This program is distributed in the hope that it will be useful,
        but WITHOUT ANY WARRANTY; without even the implied warranty of
        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
        GNU General Public License for more details.

        You should have received a copy of the GNU General Public License
        along with this program; if not, write to the Free Software
        Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

        ---

        Please do not change or remove any of the copyrights or links to web pages
        when modifying any of the files.

*/

function zhongwenDict() {
    this.loadDictionary();
}

zhongwenDict.prototype = {

    wordDict: '',
    wordIndex: '',

    grammarKeywords: {},

    fileRead: function(url, callback) {
        var req = new XMLHttpRequest();
        req.open("GET", url, true);

        req.onload = function (e) {
            if (req.readyState === 4) {
                if (req.status === 200) {
                    callback(req.responseText);
                } else {
                    console.error(req.statusText);
                }
            }
        };
        req.onerror = function (e) {
            console.error(req.statusText);
        };

        req.send(null);
    },

    find: function(data, text) {
        const tlen = text.length;
        var beg = 0;
        var end = data.length - 1;
        var i;
        var mi;
        var mis;

        while (beg < end) {
            mi = Math.floor((beg + end) / 2);
            i = data.lastIndexOf('\n', mi) + 1;

            mis = data.substr(i, tlen);
            if (text < mis) end = i - 1;
            else if (text > mis) beg = data.indexOf('\n', mi + 1) + 1;
            else return data.substring(i, data.indexOf('\n', mi + 1));
        }
        return null;
    },

    loadDictionary: function() {
        this.fileRead(chrome.extension.getURL("data/cedict_ts.u8"), function(text) { zhongwenDict.prototype.wordDict = text; });
        this.fileRead(chrome.extension.getURL("data/cedict.idx"), function(text) { zhongwenDict.prototype.wordIndex = text; });
        this.fileRead(chrome.extension.getURL("data/grammarKeywordsMin.json"),
            function(text) {
                zhongwenDict.prototype.grammarKeywords = JSON.parse(text)
            });
    },

    hasKeyword: function (keyword) {
        return this.grammarKeywords[keyword];
    },
    
    wordSearch: function(word, max) {
        var entry = { };

        var dict = this.wordDict;
        var index = this.wordIndex;
        var maxTrim = 7;
        var cache = {};
        var have = {};
        var count = 0;
        var maxLen = 0;

        if (max != null){
            maxTrim = max;
        }

        entry.data = [];

        while (word.length > 0) {
            var ix = cache[word];
            if (!ix) {
                ix = this.find(index, word + ',');
                if (!ix) {
                    cache[word] = [];
                    continue;
                }
                ix = ix.split(',');
                cache[word] = ix;
            }

            for (var j = 1; j < ix.length; ++j) {
                var offset = ix[j];
                if (have[offset]) continue;

                var dentry = dict.substring(offset, dict.indexOf('\n', offset));

                if (count >= maxTrim) {
                    entry.more = 1;
                    break;
                }

                have[offset] = 1;
                ++count;
                if (maxLen == 0) maxLen = word.length;

                entry.data.push([dentry, word]);
            } // for j < ix.length
            if (count >= maxTrim) break;
            word = word.substr(0, word.length - 1);
        } // while word.length > 0

        if (entry.data.length == 0) return null;

        entry.matchLen = maxLen;
        return entry;
    },

    translate: function(text) {
        var e, o;
        var skip;

        o = {};
        o.data = [];
        o.textLen = text.length;

        while (text.length > 0) {
            e = this.wordSearch(text, 1);
            if (e != null) {
                if (o.data.length >= 7) {
                    o.more = 1;
                    break;
                }
                o.data.push(e.data[0]);
                skip = e.matchLen;
            }
            else {
                skip = 1;
            }
            text = text.substr(skip, text.length - skip);
        }

        if (o.data.length == 0) {
            return null;
        }

        o.textLen -= text.length;
        return o;
    }
};


background.js
/*
        Zhongwen - A Chinese-English Popup Dictionary
        Copyright (C) 2012 Christian Schiller
        https://chrome.google.com/extensions/detail/kkmlkkjojmombglmlpbpapmhcaljjkde
*/

chrome.browserAction.onClicked.addListener(zhongwenMain.enableToggle);
chrome.tabs.onActiveChanged.addListener(zhongwenMain.onTabSelect);

chrome.extension.onRequest.addListener(function(request, sender, response) {

    switch(request.type) {
        case 'enable?':
            zhongwenMain.onTabSelect(sender.tab.id);
            break;
        case 'search':
            var e = zhongwenMain.search(request.text);
            response(e);
            break;
        case 'open':
                        
            var tabID = zhongwenMain.tabIDs[request.tabType];
            if (tabID) {
                chrome.tabs.get(tabID, function(tab) {
                    if (tab && (tab.url.substr(-13) == 'wordlist.html')) {
                        chrome.tabs.reload(tabID);
                        chrome.tabs.update(tabID, {
                            active: true
                        });
                    } else {
                        chrome.tabs.create({
                            url: request.url
                            }, function(tab) {
                            zhongwenMain.tabIDs[request.tabType] = tab.id;
                            if (request.tabType == 'wordlist') {
                                // make sure the table is sized correctly
                                chrome.tabs.reload(tab.id);
                            }
                        });
                    }
                });
            } else {
                chrome.tabs.create({
                    url: request.url
                    }, function(tab) {
                    zhongwenMain.tabIDs[request.tabType] = tab.id;
                    if (request.tabType == 'wordlist') {
                        // make sure the table is sized correctly
                        chrome.tabs.reload(tab.id);
                    }
                });
            }
                        
            break;

        case 'copy':
            var txt = document.createElement('textarea');
            txt.style.position = "absolute";
            txt.style.left = "-100%";
            txt.value = request.data;
            document.body.appendChild(txt);
            txt.select();
            document.execCommand('copy');
            document.body.removeChild(txt);
            break;
        case 'add':
            var json = localStorage['wordlist'];
                        
            var wordlist;
            if (json) {
                wordlist = JSON.parse(json);
            } else {
                wordlist = []
            }
                        
            for (var i in request.entries) {
                            
                var entry = {}
                entry.simplified = request.entries[i].simplified;
                entry.traditional = request.entries[i].traditional;
                entry.pinyin = request.entries[i].pinyin;
                entry.definition = request.entries[i].definition;
                            
                wordlist.push(entry);
            }                            
            localStorage['wordlist'] = JSON.stringify(wordlist);

            var tabID = zhongwenMain.tabIDs['wordlist'];
            if (tabID) {
                chrome.tabs.get(tabID, function(tab) {
                    if (tab) {
                        chrome.tabs.reload(tabID);
                    }
                });
            }

            break;
        case 'iframe':
            chrome.tabs.executeScript(sender.tab.id, {
                file: 'content.js',
                allFrames: true
            })
            break;
        default:
    // ignore
    }
});

function initStorage(key, defaultValue) {
    var currentValue = localStorage[key];
    if (!currentValue) {
        localStorage[key] = defaultValue;
    }
}

initStorage("popupcolor", "yellow");
initStorage("tonecolors", "yes");
initStorage("fontSize", "small");
initStorage("skritterTLD", "com");
initStorage("zhuyin", "no");
initStorage("grammar", "yes");

zhongwenMain.config = {};
zhongwenMain.config.css = localStorage["popupcolor"];
zhongwenMain.config.tonecolors = localStorage["tonecolors"];
zhongwenMain.config.fontSize = localStorage["fontSize"]
zhongwenMain.config.skritterTLD = localStorage.skritterTLD;
zhongwenMain.config.zhuyin = localStorage.zhuyin;
zhongwenMain.config.grammar = localStorage.grammar;
zhongwenMain.config.simpTrad = localStorage.simpTrad;

if (localStorage['enabled'] == 1) {
   zhongwenMain.loadDictionary();
   zhongwenMain.enabled = 1;
} else {
    zhongwenMain.enabled = 0;
}


js/options.js
/*
        Zhongwen - A Chinese-English Popup Dictionary
        Original Work Copyright (C) 2012 Christian Schiller
        https://chrome.google.com/extensions/detail/kkmlkkjojmombglmlpbpapmhcaljjkde
        Modified work Copyright (C) 2017 Leonard Lausen
        https://github.com/leezu/zhongwen
*/

function restoreOptions() {
  var optionsPromise = browser.storage.sync.get({
    options: {
      'popupcolor': 'yellow',
      'tonecolors': 'yes',
      'fontSize': 'small',
      'skritterTLD': 'com',
      'zhuyin': 'no',
      'grammar': 'yes',
      'simpTrad': 'classic',
    }
  });
  optionsPromise.then((storage) => {
    let options = storage.options;
    for (var i = 0; i < document.optform.popupcolor.length; i++) {
      if (document.optform.popupcolor[i].value ==
        options.popupcolor) {
        document.optform.popupcolor[i].selected = true;
        break;
      }
    }

    if (options.tonecolors == 'no') {
      document.optform.tonecolors[1].selected = true;
    } else {
      document.optform.tonecolors[0].selected = true;
    }

    if (options.fontSize == 'small') {
      document.optform.fontSize[1].selected = true;
    } else {
      document.optform.fontSize[0].selected = true;
    }

    if (options.skritterTLD == 'cn') {
      document.optform.skritterTLD[1].selected = true;
    } else {
      document.optform.skritterTLD[0].selected = true;
    }

    if (options.zhuyin == 'yes') {
      document.optform.zhuyin[1].selected = true;
    } else {
      document.optform.zhuyin[0].selected = true;
    }

    if (options.grammar == 'no') {
      document.optform.grammar[1].selected = true;
    } else {
      document.optform.grammar[0].selected = true;
    }
    if (options.simpTrad == 'auto') {
      document.optform.simpTrad[1].selected = true;
    } else {
      document.optform.simpTrad[0].selected = true;
    }
  });
}

function saveOptions() {
  let options = {
    'popupcolor': document.optform.popupcolor.value,
    'tonecolors': document.optform.tonecolors.value,
    'fontSize': document.optform.fontSize.value,
    'skritterTLD': document.optform.skritterTLD.value,
    'zhuyin': document.optform.zhuyin.value,
    'grammar': document.optform.grammar.value,
    'simpTrad': document.optform.simpTrad.value
  };
  let setting = browser.storage.sync.set({
    options
  });
}

document.addEventListener('DOMContentLoaded',
  restoreOptions);
document.querySelector("form").addEventListener("submit",
  saveOptions);


js/wordlist.js
/*
        Zhongwen - Ein Chinesisch-Deutsch Popup-Wörterbuch
        Original Work Copyright (C) 2011-2013 Christian Schiller
        https://chrome.google.com/webstore/detail/jjkbnbgakjgfiajfkifdbhbfmjgmddeh
        German version of the Chinese-English Zhongwen Popup-Dictionary
        https://chrome.google.com/webstore/detail/kkmlkkjojmombglmlpbpapmhcaljjkde
        Modified work Copyright (C) 2017 Leonard Lausen
        https://github.com/leezu/zhongwen
*/

var wlstring = localStorage['wordlist'];

var entries;
if (wlstring) {
  entries = JSON.parse(wlstring);
} else {
  entries = [];
}

$(function() {

  $('#words').jqGrid({
    data: entries,
    datatype: 'local',
    colNames: ['Simplified', 'Traditional',
      'Pinyin', 'Definition'
    ],
    colModel: [{
        name: 'simplified',
        index: 'simplified'
      },

      {
        name: 'traditional',
        index: 'traditional'
      },

      {
        name: 'pinyin',
        index: 'pinyin'
      },

      {
        name: 'definition',
        index: 'definition',
        cellattr: function(rowId, tv, rawObject,
          cm, rdata) {
          return 'style="white-space: normal;'
        }
      },
    ],
    multiselect: true,
    rowNum: 1000,
    rowList: [10, 20, 30],
    viewrecords: true,
    gridview: true,
    caption: 'Zhongwen: My Words',
    autowidth: true,
    height: '100%',
    emptyrecords: "No words",
    loadui: 'disable'
  });

  $('#savebutton').click(function(e) {
    var selected = jQuery("#words").getGridParam(
      'selarrrow');

    selected.sort();

    var set = {};
    for (var i in selected) {
      set[selected[i]] = 1;
    }

    var content = '';
    for (i in entries) {
      var entry_id = entries[i].id;
      if (set[entry_id]) {
        var entry = entries[i];
        content += entry.simplified;
        content += '\t';
        content += entry.traditional;
        content += '\t';
        content += entry.pinyin;
        content += '\t';
        content += entry.definition;
        content += '\r\n';
      }
    }

    var saveBlob = new Blob([content], {
      "type": "text/plain"
    });
    window.URL = window.URL || window.webkitURL; // Support Firefox & Chrome
    url = window.URL.createObjectURL(saveBlob);

    function onStartedDownload(id) {
      console.log(`Started downloading: ${id}`);
    }

    function onFailed(error) {
      console.log(`Download failed: ${error}`);
    }

    var downloading = browser.downloads.download({
      url: url,
      filename: "zhongwen-words.txt",
      conflictAction: 'uniquify',
      saveAs: true
    });
    downloading.then(onStartedDownload, onFailed);
  });

  $('#delete').click(function(e) {
    var selected = jQuery("#words").getGridParam(
      'selarrrow');

    selected.sort();

    var set = {};
    for (var i in selected) {
      set[selected[i]] = 1;
    }

    var toKeep = [];
    for (i in entries) {
      var entry_id = entries[i].id;
      if (!set[entry_id]) {
        toKeep.push(entries[i]);
      }
    }

    localStorage['wordlist'] = JSON.stringify(
      toKeep);

    location.reload(true);
  });

  $('#help').click(function(e) {
    $('#helpmsg').toggle('bounce', function() {
      if ($('#help').text() == 'Show help') {
        $('#help').text('Hide help');
      } else {
        $('#help').text('Show help');
      }
    });

  });

  if (entries.length > 0) {
    $('#nodata').hide();
    $('#save').show();
    $('#delete').show();
    $('#help').show();
  } else {
    $('#nodata').show();
    $('#save').hide();
    $('#delete').hide();
    $('#help').hide();
  }

});


main.js
/*
        Zhongwen - A Chinese-English Popup Dictionary
        Original Work Copyright (C) 2011 Christian Schiller
        https://chrome.google.com/extensions/detail/kkmlkkjojmombglmlpbpapmhcaljjkde
        Modified work Copyright (C) 2017 Leonard Lausen
        https://github.com/leezu/zhongwen

        ---

        Originally based on Rikaikun 0.8
        Copyright (C) 2010 Erek Speed
        http://code.google.com/p/rikaikun/

        ---

        Originally based on Rikaichan 1.07
        by Jonathan Zarate
        http://www.polarcloud.com/

        ---

        Originally based on RikaiXUL 0.4 by Todd Rudick
        http://www.rikai.com/
        http://rikaixul.mozdev.org/

        ---

        This program is free software; you can redistribute it and/or modify
        it under the terms of the GNU General Public License as published by
        the Free Software Foundation; either version 2 of the License, or
        (at your option) any later version.

        This program is distributed in the hope that it will be useful,
        but WITHOUT ANY WARRANTY; without even the implied warranty of
        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
        GNU General Public License for more details.

        You should have received a copy of the GNU General Public License
        along with this program; if not, write to the Free Software
        Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

        ---

        Please do not change or remove any of the copyrights or links to web pages
        when modifying any of the files.

*/
function reportError(error) {
  console.error(`Error: ${error}`);
}
function ignoreError(error) {}

var zhongwenMain = {

    altView: 0,

    tabIDs: {},

    loadDictionary: async function() {
      let dictData = await loadDictData();
      return new ZhongwenDictionary(...dictData);
    },

    // The callback for onActivated.
    // Just sends a message to the tab to enable itself if it hasn't already.
    onTabActivated: function (activeInfo) {
        zhongwenMain._checkEnableTab(activeInfo.tabId)
    },
    onTabUpdated: function (tabId, changeInfo, tabInfo) {
        zhongwenMain._checkEnableTab(tabId)
    },
    _checkEnableTab: function (tabId) {
        let enabledPromise = browser.storage.local.get({enabled: 0});
        enabledPromise.then((storage) => {
            if (storage.enabled === 1) {
                let optionsPromise = browser.storage.sync.get({
                    options: {
                        'hanziSimplTrad': "both",
                        'popupcolor': "yellow",
                        'tonecolors': "yes",
                        'fontSize': "small",
                        'skritterTLD': "com",
                        'zhuyin': "no",
                        'grammar': "yes"
                    }
                });
                optionsPromise.then((storage) => {
                    browser.tabs.sendMessage(tabId, {
                        type: "enable",
                        config: storage.options
                    }).catch(reportError);
                });
            }
        });
    },

  enable: function(tab) {
    let optionsPromise = browser.storage.sync.get({
      options: {
        'hanziSimplTrad': "both",
        'popupcolor': 'yellow',
        'tonecolors': 'yes',
        'fontSize': 'small',
        'skritterTLD': 'com',
        'zhuyin': 'no',
        'grammar': 'yes'
      }
    })
    let dictionaryPromise = zhongwenMain.loadDictionary()
    let enabled = 1
    let enablePromise = browser.storage.local.set({enabled})

    Promise.all([optionsPromise, dictionaryPromise, enablePromise]).then(
      ([storage, dictionary, enabled]) => {

        this.dict = dictionary

        // Send message to current tab to add listeners and create stuff
        if (tab !== undefined) {
          browser.tabs.sendMessage(tab.id, {
            type: 'enable',
            config: storage.options
          }).catch(reportError)

          browser.tabs.sendMessage(tab.id, {
            type: 'showPopup',
            isHelp: true
          }).catch(reportError)
        }

        browser.browserAction.setBadgeBackgroundColor({
          'color': [255, 0, 0, 255]
        })

        browser.browserAction.setBadgeText({
          'text': 'On'
        })

        browser.contextMenus.create({
          title: 'Open word list',
          id: 'wordlist-page',
          onclick: zhongwenMain.wordlistTab,
          contexts: ['page']
        })
      })
  },

    disable: function(tab) {
      let enabled = 0;
      let enablePromise = browser.storage.local.set({enabled});

      Promise.all([enablePromise]).then(([storage]) => {
        // Delete dictionary object after we implement it
        delete this.dict;

        browser.browserAction.setBadgeBackgroundColor({
          "color": [0, 0, 0, 0]
        });
        browser.browserAction.setBadgeText({
          "text": ""
        });

        // Send a disable message to all browsers.
        browser.windows.getAll({
          "populate": true
        }).then((windowInfoArray) => {
          for (let windowInfo of windowInfoArray) {
            for (let tabInfo of windowInfo.tabs) {
              browser.tabs.sendMessage(tabInfo.id, {
                type: "disable"
              }).catch(ignoreError);
              // Some tabs may not have a listener as they were never activated
            }
          }
        });
      });

      browser.contextMenus.remove("wordlist-page");
    },

  enableToggle: function(tab) {
    let enabledPromise = browser.storage.local.get({enabled: 0});
    enabledPromise.then((storage) => {
      if (storage.enabled == 1) {
        zhongwenMain.disable(tab);
      } else {
        zhongwenMain.enable(tab);
      }
    });
  },

    search: function(text) {

        var entry = this.dict.wordSearch(text);
        if (entry != null) {
            for (var i = 0; i < entry.data.length; i++) {
                var word = entry.data[i][1];
                if (this.dict.hasKeyword(word) && (entry.matchLen == word.length)) {
                    // the final index should be the last one with the maximum length
                    entry.grammar = { keyword: word, index: i };
                }
            }
        }

        return entry;

    },

  wordlistTab: function() {
    var url = browser.extension.getURL("/wordlist.html");
    var tabID = zhongwenMain.tabIDs['wordlist'];
    if (tabID) {
      browser.tabs.get(tabID, function(tab) {
        if (tab && (tab.url.substr(-13) == 'wordlist.html')) {
          browser.tabs.reload(tabID);
          browser.tabs.update(tabID, {active: true});
        } else {
          browser.tabs.create({
            url: url
          }, function(tab) {
            zhongwenMain.tabIDs['wordlist'] = tab.id;
            browser.tabs.reload(tab.id);
          });
        }
      });
    } else {
      browser.tabs.create({ url: url }, function(tab) {
        zhongwenMain.tabIDs['wordlist'] = tab.id;
        browser.tabs.reload(tab.id); });
    }
  }
};
